module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php

```diff
@@ -272,19 +297,36 @@ public function loadPickUpLocationsFromNcip()
     protected function sendRequest($xml)
     {
         $this->debug('Sendig NCIP request: ' . $xml);
+        $client = $this->httpService->createClient($this->url);
+        if ($this->useOAuth2) {
+            $client->getRequest()->getHeaders()
+                ->addHeaderLine('Authorization', $this->getOAuth2Token());
+        }
+        // Set timeout value
+        $timeout = $this->config['Catalog']['http_timeout'] ?? 30;
+        $client->setOptions(['timeout' => $timeout]);
+        $client->setRawBody($xml);
+        $client->setEncType('application/xml; charset=UTF-8');
+        $client->setMethod('POST');
         // Make the NCIP request:
         try {
-            $client = $this->httpService->createClient($this->url);
-            // Set timeout value
-            $timeout = $this->config['Catalog']['http_timeout'] ?? 30;
-            $client->setOptions(['timeout' => $timeout]);
-            $client->setRawBody($xml);
-            $client->setEncType('application/xml; charset=UTF-8');
-            $result = $client->setMethod('POST')->send();
+            $result = $client->send();
         } catch (\Exception $e) {
             throw new ILSException($e->getMessage());
         }
 
+        // If we get a 401, we need to renew the access token and try again
+        if ($this->useOAuth2 && $result->getStatusCode() == 401) {
+            $client->getRequest()->getHeaders()
+                ->addHeaderLine('Authorization', $this->getOAuth2Token(true));
+
+            try {
+                $result = $client->send();
+            } catch (\Exception $e) {
+                throw new ILSException($e->getMessage());
```

Comment:
I think it might be a good idea to log the actual errors and return something that doesn't potentially reveal too much information the ILSException.

Reply:
fixed